docs/basic-features/typescript.md

```diff
@@ -122,6 +122,45 @@ function MyApp({ Component, pageProps }: AppProps) {
 export default MyApp
 ```
 
+## Custom `Document`
+
+With a [custom `Document`](/docs/advanced-features/custom-document), you can use the built-in types and change the file name to `./pages/_document.tsx` like so:
+
+```tsx
+import Document, {
+  Html, Head, Main, NextScript,
+  DocumentContext,
+  DocumentInitialProps,
+  /* DocumentProps, */
+} from 'next/document'
+
+class MyDocument extends Document {
+  static async getInitialProps(ctx: DocumentContext): Promise<DocumentInitialProps> {
```

Comment:
Per my testing, only `DocumentContext` is required to have `_document.tsx` ready with types, this example feels a a little overkill, if possible we can mention `DocumentContext` only, and add it to the custom document page instead of here.

Reply:
Hey @lfades! Thanks for reviewing.

You are right, we only need `DocumentContext` is enough to get auto-completion and code check for the `ctx` variable.

`DocumentInitialProps` can be useful if you are returning more things from the `getInitialProps`, something like:

```tsx
interface MyCustomProps { hey: String }

// Then...

class MyDocument extends Document {
  static async getInitialProps(
    ctx: DocumentContext,
  ): Promise<MyCustomProps & DocumentInitialProps> {
    const initialProps = await Document.getInitialProps(ctx);
    return { ...initialProps, hey: 'from next.js' };
  }
}
```

With that we get the validation that we always return the correct values. What do you think, is this too much?

I've never used `DocumentProps` so I can't think of a reason to document it 😆. I just though it would be nice to have everything documented.

---

Also, if we move this to the Document page, should we also move the App section to the App Page to keep the consistency?

Reply:
I think `DocumentContext` is enough 👍 

The App section can stay as it's right now, I have some plans for the future about documentation in TypeScript that may improve all of this 🕵️

Reply:
Awesome, looking forward to these improvements! :relaxed:

Made the requested changes. If there's anything else I can do let me know.